pkg/policy/policy.go

```diff
@@ -0,0 +1,135 @@
+	"knative.dev/pkg/apis"
+)
+
+type Verification struct {
+	// NoMatchPolicy specifies the behavior when a base image doesn't match any
+	// of the listed policies.  It allows the values: allow, deny, and warn.
+	NoMatchPolicy string `yaml:"no-match-policy,omitempty"`
+
+	// Policies specifies a collection of policies to use to cover the base
+	// images used as part of evaluation.  See "policy" below for usage.
+	// Policies can be nil so that we can distinguish between an explicitly
+	// specified empty list and when policies is unspecified.
+	Policies *[]Source `yaml:"policies,omitempty"`
+}
+
+// Source contains a set of options for specifying policies.  Exactly
+// one of the fields may be specified for each Source entry.
+type Source struct {
+	// Data is a collection of one or more ClusterImagePolicy resources.
+	Data string `yaml:"data,omitempty"`
+
+	// Path is a path to a file containing one or more ClusterImagePolicy
+	// resources.
+	Path string `yaml:"path,omitempty"`
+
+	// URL links to a file containing one or more ClusterImagePolicy resources.
+	URL string `yaml:"url,omitempty"`
+}
+
+func (v *Verification) Validate(ctx context.Context) (errs *apis.FieldError) {
+	switch v.NoMatchPolicy {
+	case "allow", "deny", "warn":
+		// Good!
+	case "":
+		errs = errs.Also(apis.ErrMissingField("noMatchPolicy"))
+	default:
+		errs = errs.Also(apis.ErrInvalidValue(v.NoMatchPolicy, "noMatchPolicy"))
+	}
+
+	if v.Policies == nil {
+		errs = errs.Also(apis.ErrMissingField("policies"))
+	} else {
+		for i, p := range *v.Policies {
+			errs = errs.Also(p.Validate(ctx).ViaFieldIndex("policies", i))
+		}
+	}
+
+	return errs
+}
+
+func (pd *Source) Validate(ctx context.Context) (errs *apis.FieldError) {
+	// Check that exactly one of the fields is set.
+	set := sets.NewString()
+	if pd.Data != "" {
+		set.Insert("data")
+		_, _, err := ParseClusterImagePolicies(ctx, pd.Data)
+		if err != nil {
+			errs = errs.Also(apis.ErrInvalidValue(err.Error(), "data"))
+		}
+	}
+	if pd.Path != "" {
+		set.Insert("path")
+		raw, err := os.ReadFile(pd.Path)
+		if err != nil {
+			errs = errs.Also(&apis.FieldError{
+				Message: err.Error(),
+				Paths:   []string{"path"},
+			})
+		} else {
+			_, _, err := ParseClusterImagePolicies(ctx, string(raw))
+			if err != nil {
+				errs = errs.Also(apis.ErrInvalidValue(err.Error(), "path"))
+			}
+		}
+	}
+	if pd.URL != "" {
+		set.Insert("url")
+		resp, err := http.Get(pd.URL)
+		if err != nil {
+			errs = errs.Also(&apis.FieldError{
+				Message: err.Error(),
+				Paths:   []string{"url"},
+			})
+		} else {
+			defer resp.Body.Close()
+			raw, err := io.ReadAll(resp.Body)
+			if err != nil {
+				errs = errs.Also(&apis.FieldError{
+					Message: err.Error(),
+					Paths:   []string{"url"},
+				})
+			} else {
+				_, _, err := ParseClusterImagePolicies(ctx, string(raw))
+				if err != nil {
+					errs = errs.Also(apis.ErrInvalidValue(err.Error(), "url"))
+				}
+			}
+		}
+	}
+
+	switch set.Len() {
```

Comment:
This is bit clunky because if somebody specifies multiple values, we'll try to fetch/parse them, and after that will then (in addition to those errors) print this. I think it would be better to first see if the 'spec' is valid and only if it's valid (no multiples, or missing) proceed to fetch/validate them.

Reply:
Just a thought, I wonder if instead of fetching/validating in each of the if clauses above, you just fetch the bytes into something like:
```
var raw string
var field string
if pd.Data != "" {
  raw = pd.Data
  field = "data"
}

if pd.Path != "" {
  if field != "" {
      // multiple paths specified, error out
  }
field = "path"
  raw, err = os.ReadFile(...
}
```

And then after the if clauses you'd just have the now duplicated method that deals with raw alone?
```
		_, _, err := ParseClusterImagePolicies(ctx, raw)
		if err != nil {
			errs = errs.Also(apis.ErrInvalidValue(err.Error(), path))
		}
```

Reply:
Yeah, I was just trying to avoid duplicating the `if` cases, I can rework this in the next push.